types/set.go

```diff
@@ -0,0 +1,276 @@
+func (t SetType) ValueFromTerraform(ctx context.Context, in tftypes.Value) (attr.Value, error) {
+	if !in.Type().Is(t.TerraformType(ctx)) {
+		return nil, fmt.Errorf("can't use %s as value of Set with ElementType %T, can only use %s values", in.String(), t.ElemType, t.ElemType.TerraformType(ctx).String())
+	}
+	set := Set{
+		ElemType: t.ElemType,
+	}
+	if !in.IsKnown() {
+		set.Unknown = true
+		return set, nil
+	}
+	if in.IsNull() {
+		set.Null = true
+		return set, nil
+	}
+	val := []tftypes.Value{}
+	err := in.As(&val)
+	if err != nil {
+		return nil, err
+	}
+	elems := make([]attr.Value, 0, len(val))
+	for _, elem := range val {
+		av, err := t.ElemType.ValueFromTerraform(ctx, elem)
+		if err != nil {
+			return nil, err
+		}
+		elems = append(elems, av)
+	}
+	set.Elems = elems
+	return set, nil
+}
+
+// Equal returns true if `o` is also a SetType and has the same ElemType.
+func (t SetType) Equal(o attr.Type) bool {
+	if t.ElemType == nil {
+		return false
+	}
+	other, ok := o.(SetType)
+	if !ok {
+		return false
+	}
+	return t.ElemType.Equal(other.ElemType)
+}
+
+// ApplyTerraform5AttributePathStep applies the given AttributePathStep to the
+// set.
+func (t SetType) ApplyTerraform5AttributePathStep(step tftypes.AttributePathStep) (interface{}, error) {
+	if _, ok := step.(tftypes.ElementKeyValue); !ok {
+		return nil, fmt.Errorf("cannot apply step %T to SetType", step)
+	}
+
+	return t.ElemType, nil
+}
+
+// Validate implements type validation. This type requires all elements to be
+// unique.
+func (s SetType) Validate(ctx context.Context, in tftypes.Value, path *tftypes.AttributePath) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	if !in.Type().Is(tftypes.Set{}) {
+		err := fmt.Errorf("expected Set value, received %T with value: %v", in, in)
+		diags.AddAttributeError(
+			path,
+			"Set Type Validation Error",
+			"An unexpected error was encountered trying to validate an attribute value. This is always an error in the provider. Please report the following to the provider developer:\n\n"+err.Error(),
+		)
+		return diags
+	}
+
+	if !in.IsKnown() {
+		return diags
+	}
+
+	var elems []tftypes.Value
+
+	if err := in.As(&elems); err != nil {
+		diags.AddAttributeError(
+			path,
+			"Set Type Validation Error",
+			"An unexpected error was encountered trying to validate an attribute value. This is always an error in the provider. Please report the following to the provider developer:\n\n"+err.Error(),
+		)
+		return diags
+	}
+
+	// Attempting to use map[tftypes.Value]struct{} for duplicate detection yields:
+	//   panic: runtime error: hash of unhashable type tftypes.primitive
+	// Instead, use for loops.
+	for indexOuter, elemOuter := range elems {
+		// Only evaluate fully known values for duplicates.
+		if !elemOuter.IsFullyKnown() {
+			continue
+		}
+
+		for indexInner, elemInner := range elems {
+			if indexInner == indexOuter {
+				continue
+			}
+
+			if !elemInner.Equal(elemOuter) {
+				continue
+			}
```

Comment:
This is going to compare each value to each other _twice_. For example, when `indexOuter` is `2` and `indexInner` is `6`, they'll be compared, and when `indexOuter` is `6` and `indexInner` is `2`, they'll be compared.

Could we instead _start_ `indexOuter` at `indexInner+1`, on the understanding the element at `indexOuter` has already been compared to each element before it, when _they_ were `indexOuter`? That may help the quadratic nature of this function. (I don't know that it's a _big_ deal--we should be working with small enough numbers here, in the thousands, that I think even a quadratic function is probably fine, but thought I'd raise it.)

Reply:
I added benchmark tests for this method:

```go
func benchmarkSetTypeValidate(b *testing.B, elementCount int) {
	elements := make([]tftypes.Value, 0, elementCount)

	for idx := range elements {
		elements[idx] = tftypes.NewValue(tftypes.String, strconv.Itoa(idx))
	}

	ctx := context.Background()
	in := tftypes.NewValue(
		tftypes.Set{
			ElementType: tftypes.String,
		},
		elements,
	)
	path := tftypes.NewAttributePath().WithAttributeName("test")
	set := SetType{}

	for n := 0; n < b.N; n++ {
		set.Validate(ctx, in, path)
	}
}

func BenchmarkSetTypeValidate10(b *testing.B) {
	benchmarkSetTypeValidate(b, 10)
}

func BenchmarkSetTypeValidate100(b *testing.B) {
	benchmarkSetTypeValidate(b, 100)
}

func BenchmarkSetTypeValidate1000(b *testing.B) {
	benchmarkSetTypeValidate(b, 1000)
}

func BenchmarkSetTypeValidate10000(b *testing.B) {
	benchmarkSetTypeValidate(b, 10000)
}

func BenchmarkSetTypeValidate100000(b *testing.B) {
	benchmarkSetTypeValidate(b, 100000)
}

func BenchmarkSetTypeValidate1000000(b *testing.B) {
	benchmarkSetTypeValidate(b, 1000000)
}
```

The inner loop optimization seems like a good one, so adjusted the implementation:

```go
	for indexOuter, elemOuter := range elems {
		// Only evaluate fully known values for duplicates.
		if !elemOuter.IsFullyKnown() {
			continue
		}

		for indexInner := indexOuter + 1; indexInner < len(elems); indexInner++ {
			elemInner := elems[indexInner]

			if !elemInner.Equal(elemOuter) {
				continue
			}

			diags.AddAttributeError(
				path.WithElementKeyValue(elemInner),
				"Duplicate Set Element",
				fmt.Sprintf("This attribute contains duplicate values of: %s", elemInner),
			)
		}
	}
```

Before and after code change:

```console
$ go test -bench=. -benchtime=10s ./types
goos: darwin
goarch: amd64
pkg: github.com/hashicorp/terraform-plugin-framework/types
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkSetTypeValidate10-16           123664647               97.27 ns/op
BenchmarkSetTypeValidate100-16          123715761               96.50 ns/op
BenchmarkSetTypeValidate1000-16         124075330               97.07 ns/op
BenchmarkSetTypeValidate10000-16        123091288               97.36 ns/op
BenchmarkSetTypeValidate100000-16       120266646               99.68 ns/op
BenchmarkSetTypeValidate1000000-16      124239894               94.87 ns/op
PASS
ok      github.com/hashicorp/terraform-plugin-framework/types   131.062s
```

```console
$ go test -bench=. -benchtime=10s ./types
goos: darwin
goarch: amd64
pkg: github.com/hashicorp/terraform-plugin-framework/types
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkSetTypeValidate10-16           124338459               95.79 ns/op
BenchmarkSetTypeValidate100-16          124734000               95.74 ns/op
BenchmarkSetTypeValidate1000-16         124954414               97.24 ns/op
BenchmarkSetTypeValidate10000-16        122782779               97.07 ns/op
BenchmarkSetTypeValidate100000-16       120025662               99.47 ns/op
BenchmarkSetTypeValidate1000000-16      126757616               94.44 ns/op
PASS
ok      github.com/hashicorp/terraform-plugin-framework/types   131.266s
```

Reply:
Those sweet, sweet nanoseconds, though! :D Yeah, that jibes with my suspicions.